atlas_core/hdf_to_postgres.py

```diff
@@ -0,0 +1,175 @@
+                    file_name,
+                    hdf_tables,
+                    defer_sql_objs=True,
+                    sql_table=sql_table,
+                    csv_chunksize=csv_chunksize,
+                    hdf_chunksize=hdf_chunksize,
+                    hdf_metadata=hdf_meta,
+                )
+            )
+        elif any("partner" in table for table in hdf_tables):
+            partners.append(
+                BigHDFTableCopy(
+                    file_name,
+                    hdf_tables,
+                    defer_sql_objs=True,
+                    sql_table=sql_table,
+                    csv_chunksize=csv_chunksize,
+                    hdf_chunksize=hdf_chunksize,
+                    hdf_metadata=hdf_meta,
+                )
+            )
+        else:
+            other.append(
+                HDFTableCopy(
+                    file_name,
+                    hdf_tables,
+                    defer_sql_objs=True,
+                    sql_table=sql_table,
+                    csv_chunksize=csv_chunksize,
+                    hdf_chunksize=hdf_chunksize,
+                    hdf_metadata=hdf_meta,
+                )
+            )
+
+    # Return the objects sorted classifications, then partner, then other
+    return classifications, partners + other
+
+
+def hdf_to_postgres(
+    file_name="./data.h5",
+    keys=[],
+    processes=4,
+    engine_args=[],  # [db.engine.url]
+    engine_kwargs={},
+    maintenance_work_mem="1GB",
+    hdf_chunksize: int = 10 ** 7,
+    csv_chunksize: int = 10 ** 6,
+):
+
+    sql_to_hdf, metadata_vars = hdf_metadata(
+        file_name, keys=keys, metadata_attr="atlas_metadata", metadata_keys=["levels"]
+    )
+
+    classifications, tables = create_table_objects(
+        file_name,
+        sql_to_hdf,
+        csv_chunksize=csv_chunksize,
+        hdf_chunksize=hdf_chunksize,
+        hdf_meta=metadata_vars,
+    )
+
+    # Copy classifications first, not multiprocessed
+    for class_table in classifications:
+        # Order of data_formatters list matters here
+        copy_worker(
+            class_table,
+            engine_args,
+            engine_kwargs,
+            maintenance_work_mem=maintenance_work_mem,
+            data_formatters=[classification_to_pandas, cast_pandas],
+        )
+
+    # Use multiprocessing for the larger tables once classifications are complete
+    pool_args = zip(
+        tables,
+        [engine_args] * len(tables),
+        [engine_kwargs] * len(tables),
+        [maintenance_work_mem] * len(tables),
+        [[add_level_metadata, cast_pandas]] * len(tables),
+    )
+    try:
+        p = Pool(processes)
+        result = p.starmap_async(copy_worker, pool_args, chunksize=1)
+    finally:
+        del tables
+        p.close()
+        p.join()
+    if not result.successful():
+        # If there's an exception, throw it, but we don't care about the
+        # results
+        result.get()
+
+
+def multiload(app):
```

Comment:
I'm not sure if this should live in atlas_core or the main repo - since it's "app" related it feels like a main repo thing, but then it could probably get reused in other projects so perhaps this is better - I leave it up to you.